dartagnan/src/main/java/com/dat3m/dartagnan/wmm/utils/graph/immutable/ImmutableMapEventGraph.java

```diff
@@ -0,0 +1,225 @@
+package com.dat3m.dartagnan.wmm.utils.graph.immutable;
+
+import com.dat3m.dartagnan.program.event.Event;
+import com.dat3m.dartagnan.wmm.utils.graph.EventGraph;
+import com.dat3m.dartagnan.wmm.utils.graph.mutable.MapEventGraph;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Sets;
+
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.BiConsumer;
+import java.util.function.BiPredicate;
+import java.util.stream.Collectors;
+
+public class ImmutableMapEventGraph implements ImmutableEventGraph {
+
+    private final Map<Event, Set<Event>> data;
+    private final int size;
+    private ImmutableMapEventGraph inverse;
+
+    public ImmutableMapEventGraph(Map<Event, Set<Event>> data) {
+        this(data, data.values().stream().map(Set::size).reduce(0, Integer::sum));
+    }
+
+    private ImmutableMapEventGraph(Map<Event, Set<Event>> data, int size) {
+        this.data = data.entrySet().stream()
+                .filter(e -> !e.getValue().isEmpty())
+                .collect(ImmutableMap.toImmutableMap(
+                        Map.Entry::getKey,
+                        entry -> entry.getValue() instanceof ImmutableSet
+                                ? entry.getValue()
+                                : ImmutableSet.copyOf(entry.getValue())));
```

Comment:
`ImmutableSet.copyOf()` already recognizes if it copies another immutable set and then skips the copy.